src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/nodes/interrupts/CheckForInterruptsFullNode.java

```diff
@@ -36,24 +37,48 @@ public void execute(final VirtualFrame frame) {
         if (istate.shouldSkip()) {
             return;
         }
+        ProcessSwitch pendingSwitch = null;
         if (istate.tryInterruptPending()) {
-            signalSemaporeNode.executeSignal(frame, this, specialObjects[SPECIAL_OBJECT.THE_INTERRUPT_SEMAPHORE]);
+            try {
+                signalSemaporeNode.executeSignal(frame, this, specialObjects[SPECIAL_OBJECT.THE_INTERRUPT_SEMAPHORE]);
+            } catch (final ProcessSwitch ps) {
+                pendingSwitch = ps;
+            }
         }
         if (istate.tryWakeUpTickTrigger()) {
-            signalSemaporeNode.executeSignal(frame, this, specialObjects[SPECIAL_OBJECT.THE_TIMER_SEMAPHORE]);
+            try {
+                signalSemaporeNode.executeSignal(frame, this, specialObjects[SPECIAL_OBJECT.THE_TIMER_SEMAPHORE]);
+            } catch (final ProcessSwitch ps) {
+                pendingSwitch = ps;
+            }
         }
         if (istate.tryPendingFinalizations()) {
-            signalSemaporeNode.executeSignal(frame, this, specialObjects[SPECIAL_OBJECT.THE_FINALIZATION_SEMAPHORE]);
+            try {
+                signalSemaporeNode.executeSignal(frame, this, specialObjects[SPECIAL_OBJECT.THE_FINALIZATION_SEMAPHORE]);
+            } catch (final ProcessSwitch ps) {
+                pendingSwitch = ps;
+            }
         }
         if (istate.trySemaphoresToSignal()) {
             final ArrayObject externalObjects = (ArrayObject) specialObjects[SPECIAL_OBJECT.EXTERNAL_OBJECTS_ARRAY];
             if (!externalObjects.isEmptyType()) { // signal external semaphores
                 final Object[] semaphores = externalObjects.getObjectStorage();
                 Integer semaIndex;
                 while ((semaIndex = istate.nextSemaphoreToSignal()) != null) {
-                    signalSemaporeNode.executeSignal(frame, this, semaphores[semaIndex - 1]);
+                    try {
+                        signalSemaporeNode.executeSignal(frame, this, semaphores[semaIndex - 1]);
+                    } catch (final ProcessSwitch ps) {
+                        pendingSwitch = ps;
```

Comment:
Catching `ProcessSwitch` is a neat trick to achieve the right behavior. However, in the worst case, a lot of exceptions are thrown and given the check is supposed to be as cheap as possible, I think it'd make sense if we could change the implementation, so that the new `activeContext` is returned rather than thrown as part of a `ProcessSwitch`.

Is this something you'd like to look into, or should I take a look?

Reply:
I will take a look...